hail/src/main/scala/is/hail/io/TextMatrixReader.scala

```diff
@@ -438,225 +437,233 @@ class CompiledLineParser(
+    (c - const('0')).toI
+  }
 
-  private[this] def endField(p: Code[Int]): Code[Boolean] =
-    Code.memoize(p, "clp_end_field_p") { p =>
-      p.ceq(line.length()) || line(p).ceq(const(separator))
-    }
+  private[this] def endField(cb: EmitCodeBuilder, p: Value[Int]): Code[Boolean] = {
+    p.ceq(line.length()) || line(p).ceq(const(separator))
+  }
 
-  private[this] def endField(): Code[Boolean] =
-    endField(pos)
+  private[this] def endField(cb: EmitCodeBuilder): Code[Boolean] =
+    endField(cb, pos)
 
-  private[this] def missingOr(
-    mb: EmitMethodBuilder[_],
-    srvb: StagedRegionValueBuilder,
-    parse: Code[Unit]
-  ): Code[Unit] = {
+  private[this] def parseOptionalValue(
+    cb: EmitCodeBuilder,
+    parse: EmitCodeBuilder => PCode
+  ): IEmitCode = {
     assert(missingValue.size > 0)
-    val end = mb.genFieldThisRef[Int]()
-    val condition = Code(
-      end := pos + missingValue.size,
-      end <= line.length && endField(end) &&
-      line.invoke[Int, String, Int, Int, Boolean]("regionMatches",
-        pos, missingValue, 0, missingValue.size))
-    condition.mux(
-      Code(
-        pos := end,
-        srvb.setMissing()),
-      parse)
+    val end = cb.newLocal[Int]("parse_optional_value_end", pos + missingValue.size)
+
+    val Ldefined1 = CodeLabel()
+    val Lmissing = CodeLabel()
+
+    cb.ifx(end <= line.length,
+      cb.ifx(endField(cb, end),
+        cb.ifx(line.invoke[Int, String, Int, Int, Boolean]("regionMatches",
+          pos, missingValue, 0, missingValue.size),
+          {
+            cb.assign(pos, end)
+            cb.goto(Lmissing)
+          },
+          cb.goto(Ldefined1)
+        ),
+        cb.goto(Ldefined1)),
+      cb.goto(Ldefined1))
+
+    cb.define(Ldefined1)
+    val pc = parse(cb)
+    val Ldefined2 = CodeLabel()
+    cb.goto(Ldefined2)
+
+    IEmitCode(Lmissing, Ldefined2, pc)
   }
 
-  private[this] def skipMissingOr(mb: EmitMethodBuilder[_], skip: Code[Unit]): Code[Unit] = {
+  private[this] def skipOptionalValue(cb: EmitCodeBuilder, skip: EmitCodeBuilder => Unit): Unit = {
     assert(missingValue.size > 0)
-    val end = mb.genFieldThisRef[Int]()
-    val condition = Code(
-      end := pos + missingValue.size,
-      end < line.length && endField(end) &&
-      line.invoke[Int, String, Int, Int, Boolean]("regionMatches",
-        pos, missingValue, 0, missingValue.size))
-    condition.mux(
-      pos := end,
-      skip)
+    val end = cb.newLocal[Int]("skip_optional_value_end", pos + missingValue.size)
+
+    val Ldefined = CodeLabel()
+    val Lfinished = CodeLabel()
+
+    cb.ifx(end <= line.length,
+      cb.ifx(endField(cb, end),
+        cb.ifx(line.invoke[Int, String, Int, Int, Boolean]("regionMatches",
+          pos, missingValue, 0, missingValue.size),
+          {
+            cb.assign(pos, end)
+            cb.goto(Lfinished)
+          },
+          cb.goto(Ldefined)
+        ),
+        cb.goto(Ldefined)),
+      cb.goto(Ldefined))
```

Comment:
Why is the Ldefined lable needed? doesn't control just fall through?

Reply:
much nicer, thanks!